src/history/hs.c

```diff
@@ -1088,7 +1099,16 @@ __wt_hs_cursor_position(WT_SESSION_IMPL *session, WT_CURSOR *cursor, uint32_t bt
             WT_STAT_DATA_INCR(session, cursor_skip_hs_cur_position);
 
             WT_ERR(__wt_compare(session, NULL, &cursor->key, srch_key, &cmp));
-            if (cmp <= 0)
+            WT_ERR(cursor->get_key(cursor, &hs_btree_id, &hs_key, &hs_start_ts, &hs_counter));
+
+            /* Stop before crossing over to the next btree */
+            if (hs_btree_id != btree_id) {
+                ret = WT_NOTFOUND;
+                break;
+            }
+            if (cmp <= 0 &&
+              !__wt_txn_visible_all(
+                  session, cbt->upd_value->tw.stop_txn, cbt->upd_value->tw.durable_stop_ts))
```

Comment:
Could we structure this as `if (__wt_txn_visible_all(...)) { continue; }`, and move it further up before the compare and the get_key? That should be a cheaper operation.

Reply:
If you did that you run the risk of doing the full history store traversal, no?